src/java/org/apache/cassandra/service/ReplicaFilteringProtection.java

```diff
@@ -212,83 +156,122 @@ private UnfilteredPartitionIterator executeReadCommand(ReadCommand cmd, InetAddr
      * <p>
      * The listener will track both the accepted data and the primary keys of the rows that are considered as outdated.
      * That way, once the query results would have been merged using this listener, further calls to
-     * {@link #queryProtectedPartitions(int)} will use the collected data to return a copy of the
+     * {@link #queryProtectedPartitions(PartitionIterator, int)} will use the collected data to return a copy of the
      * data originally collected from the specified replica, completed with the potentially outdated rows.
      */
     UnfilteredPartitionIterators.MergeListener mergeController()
     {
-        return (partitionKey, versions) -> {
+        return new UnfilteredPartitionIterators.MergeListener()
+        {
+            @Override
+            public void close()
+            {
+                // If we hit the failure threshold before consuming a single partition, record the current rows cached.
+                tableMetrics.rfpRowsCachedPerQuery.update(Math.max(currentRowsCached, maxRowsCached));
```

Comment:
I think it would be slightly less efficient but slightly easier to read and a bit less error prone if we just updated `maxRowsCached` when we increase `currentRowsCached`, so `maxRowsCached` is always updated. Not saying that we should do it, I'm not sure if doing so would worth it.

Reply:
I'm comfortable with avoiding a potentially huge number of updates to `maxRowsCached` w/ a large partition, given the comment here is pretty clear.

Reply:
Works for me.